File: Problems/0142. (M) Linked List Cycle II/Java/solution.java
public class Solution {
  public ListNode detectCycle(ListNode head) {
    // two pointers

    // sanity check
    if (head == null || head.next == null)
      return null;

    // cycle detection
    ListNode slow = head.next;
    ListNode fast = head.next != null? head.next.next : null;
    int slow_d = 1;
    while (slow != fast) {
      if (slow == null || fast == null)
        return null;
      slow = slow.next;
      fast = fast.next != null? fast.next.next : null;
      slow_d++;
    }

    // tail.next search
    slow = head;
    fast = head;
    for (int i = 0; i < slow_d; i++)
      fast = fast.next;
    while (slow != fast) {
      slow = slow.next;
      fast = fast.next;
    }
    return slow;
  }
}

File: Problems/0142. (M) Linked List Cycle II/Python3/solution.py
class Solution:
  def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
    # two pointers

    # sanity check
    if not head or not head.next:
      return None
    
    # cycle detection
    slow = head.next
    fast = head.next.next
    slow_d = 1
    while slow != fast:
      if not slow or not fast:
        return None
      slow = slow.next
      fast = fast.next.next if fast.next else None
      slow_d += 1
    
    # tail.next search
    slow = fast = head
    for i in range(slow_d):
      fast = fast.next
    while (slow != fast):
      slow = slow.next
      fast = fast.next

    return slow

